Run operations new

Ovec8hkin opened the pull request:
run_operations takes a google sheet id as input, and performs the following routine:

* generates templates for all datasets in sheet
* determines whether there is new data available for the dataset via `ssara_federated_query-cj`
* runs `process_rsmas` on the dataset if new data is available
* moves all output files to the OPERATIONS directory

Ovec8hkin committed:
Created argument parser and associated command line options

Ovec8hkin committed:
Moved `check_for_operations_directories_and_initiate.py` to initiate_operations() function

Ovec8hkin committed:
Generated templates for necessary datasets with appropriate options

Ovec8hkin committed:
Obtains datasets to process from list of accessible template files

Ovec8hkin committed:
Merge branch 'master' into run_operations_new

# Conflicts:
#	run_operations.py

Ovec8hkin committed:
Makes call to download_ssara_rsmas.py to download data

Ovec8hkin committed:
Obtains date of newest data from `ssara_federated_query.py` and date of most recent successful data download from `stored_date.date` file.

Ovec8hkin committed:
Runs process_rsmas.py when the most recent data is more current than the last time data was downloaded.

Ovec8hkin committed:
Waits for all output files to appear and copies them over to the appropriate OPERATIONS directory

Ovec8hkin committed:
Added comments

Ovec8hkin committed:
Merge branch 'master' into run_operations_new

# Conflicts:
#	run_operations.py

Ovec8hkin committed:
Merged

Ovec8hkin committed:
Decoded subprocess outputs and changed --bsub options to --submit

Ovec8hkin committed:
Merged

Ovec8hkin committed:
generate_template_files returns list of template files generated

Rather than globbing all of the template files in the $OPERATIONS/TEMPLATES directory, run_operations now obtains a list of the template files generated by generate_template_files directly.

Ovec8hkin committed:
Made process_rsmas subprocess call use shell=True

Ovec8hkin committed:
Changed dataset_template to use custom_template_file parameters instead of working directory template file

Ovec8hkin committed:
generate_tenplate_files only returns the list of tenplates for which process_flag is TRUE instead of all datasets in the csv file

Ovec8hkin committed:
Removes file if file with same name as inps.work_dir exists before creating inps.work_dir

Ovec8hkin committed:
Add logging statements for each step of process_rsmas foe asier debugging

Ovec8hkin committed:
Removed sys.exit call in create_batch to prevent percolation of error up to subprocess in run_operations

Ovec8hkin committed:
Added proper logging to run_operations.py

Ovec8hkin committed:
Implemented writing to `stored_date.date` file for new dataset

In the case that the dataset being processed has not been previously processed, its download date is still added to the 'stored_date.date' file.

Ovec8hkin committed:
Added 'inps.template' to recursive run_ssara command

Ovec8hkin committed:
Updated logging commands across relevent run_operations scripts

Ovec8hkin committed:
run_operations.py updates

Ovec8hkin committed:
Renamed process_rsmas_options array

Ovec8hkin committed:
Changed `generate_templates_with_options` to take specific inps
variables rather than the entire inps namespace dictionary

Ovec8hkin committed:
`run_operations` Now Ingests Multiple Sheet Files

To allow for separation of datasets into multiple google sheets based on
project, and to run `run_operations.py` on all of the those datasets at
the same time, multiple sheets files can be specified, via their
sheet_id over the command line, using multiple `--sheet_id` parameters.

Ovec8hkin committed:
Standardized means of waiting for the existence of output files

`run_operations` now uses the same syntax and logic to wait for the
existence of all output files as `create_batch`.

Ovec8hkin committed:
Simplified `overwrite_stored_date`

Ovec8hkin committed:
Modified process_rsmas output job file names

Ovec8hkin committed:
Merge remote-tracking branch 'origin/run_operations_new' into run_operations_new

Ovec8hkin committed:
Merge conflicts

Ovec8hkin committed:
Modified get_datasets_to_process to use template_files returned by generation_templates rather than all templates in the /nethome/jaz101/insarlab/OPERATIONS/TEMPLATES directory

Ovec8hkin committed:
Fixed date parsing from using \n in date string

Ovec8hkin committed:
Added in work_dir creation to create_batch in case the required directoried are not already present

Ovec8hkin committed:
Added --operations option to process_rsmas

Ovec8hkin committed:
Added overwrite_stored_date_file to processSteps and associated call in
process_rsmas

Ovec8hkin committed:
Modifed cb.submit_script() to return job number for use by run_operations

Ovec8hkin committed:
Modified overwrite_stored_date_file() to propeerly write information to disk

Ovec8hkin committed:
Improved process_rsmas logging

Ovec8hkin committed:
Modified copy_output_file parameters and routine

Ovec8hkin committed:
Changed log file to be written to /nethome/jaz101/insarlab/OPERATIONS directory

gravelcycles commented on _processSteps.py:
```diff
@@ -473,3 +481,30 @@ def run_ingest_insarmaps(inps):
+###############################################################################
+
+
+def overwrite_stored_date(dset):
```
Is this function correct? I don't see how the updated lines are ever written to the file.

Something like [this](https://www.google.com/search?q=overwrite+specific+line+in+file+python&oq=overwrite+specific+line+in+file+python&aqs=chrome..69i57j0l3.9547j1j7&sourceid=chrome&ie=UTF-8) seems a bit cleaner, but if there are multiple versions of this same program running, it could lead to concurrency issues.

Ovec8hkin replied:
Hmm. Thats odd. I thought I removed that function. You are correct in that it doesn't work properly (ie. not writing new data). I was messing around with whether it was better to include at the end of `process_rsmas` itself or let `run_operations` handle it. I must have missed remove the function from the final push.

Ovec8hkin committed:
Merge remote-tracking branch 'origin/run_operations_new' into run_operations_new

Ovec8hkin committed:
Removed calls to overwrite_date_file as they are handled in run_operations

Ovec8hkin commented on _processSteps.py:
```diff
@@ -14,6 +14,8 @@
 import shutil
 import subprocess
+from datetime import datetime
+import time
```
Remove these import statements

Ovec8hkin commented on _processSteps.py:
```diff
@@ -473,3 +475,4 @@ def run_ingest_insarmaps(inps):
 
     if inps.stopinsarmaps:
         logger.log(loglevel.DEBUG, 'Exit as planned after insarmaps')
+
```
Remove

Ovec8hkin commented on create_batch.py:
```diff
@@ -268,11 +271,16 @@ def submit_script(job_name, job_file_name, argv, work_dir, walltime, email_notif
     write_single_job_file(job_name, job_file_name, command_line, work_dir, email_notif,
                           walltime=walltime, queue=os.getenv("QUEUENAME"))
-    submit_single_job("{0}.job".format(job_file_name), work_dir)
+    return submit_single_job("{0}.job".format(job_file_name), work_dir)
```
Add to documentation about return value

Ovec8hkin commented on download_ssara_rsmas.py:
```diff
@@ -143,7 +143,7 @@ def run_ssara(inps, run_number=1):
            logger.log(loglevel.WARNING, "Hanging, running again")
 
-        run_ssara(inps, run_number=run_number + 1)
+        run_ssara(inps.template, run_number=run_number + 1)
```
Change this from inps.template to template

Ovec8hkin commented on generate_template_files.py:
```diff
@@ -160,7 +160,7 @@ def generate_template_files(df, dataset):
             file_base = col_name
             output_files[file_base] = generate_template_file(names, subnames, list(df[col_name]), list(df['Comments']))
-    return output_files
+    return {k:v for (k, v) in output_files.items() if v is not None}
```
Add comment for clarity

Ovec8hkin replied:
Remove if v is none case on 183-184

Ovec8hkin replied:
Remove `generate_and_save_template_files_from_dataframe`

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
 import subprocess
 import argparse
-from datetime import datetime
+from datetime import date, datetime
```
Remove import date

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+    newest_data = ssara_output.decode('utf-8').split("\n")[-2]
+    date = datetime.strptime(newest_data.split(",")[3], DATE_FORMAT)
+
+    return date
```
Return `datetime.strptime(newest_data.split(",")[3], DATE_FORMAT)`

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
-################### directory initiation for Josh to review (and remove comment)
-##################
-
 import generate_template_files as gt
```
Don't rename to `gt`

gravelcycles replied:
Easier to read and understand code

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
-
 import generate_template_files as gt
+from rsmas_logging import RsmasLogger, loglevel
+from dataset_template import Template
```
just import `dataset_template`

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+STORED_DATE_FILE = OPERATIONS_DIRECTORY + "/stored_date.date"
+TEMPLATE_DIRECTORY = OPERATIONS_DIRECTORY + "/TEMPLATES"
+DATE_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
 
```
Move "ERRORS_DIRECTORY", "LOGS_DIRECTORY" and delete lower case variables in `initiate`

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
 
-logger = logging.getLogger()
-logger.setLevel(logging.DEBUG)
+def create_process_rsmas_parser():
```
Move parser to end of file. Move most relevant code to top of file for easing understanding.

Ovec8hkin replied:
@2gotgrossman Im going to leave this for a different diff as it should be done across the entire codebase at once

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+def get_last_downloaded_date(dset):
+
+    dataset_line = None
+    with open(STORED_DATE_FILE) as date_file:
```
Add 'r' to clarify read only option

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+    
+    job_number = process_rsmas.split('\n')[0]
+
+    stdout_file = "{}/{}/process_rsmas_{}.o".format(os.getenv('SCRATCHDIR'), dataset, job_number)
```
Move `os.getenv('SCRATCHDIR')` and `os.getenv('OPERATIONS')` to global constants

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+    logger_run_operations.log(loglevel.INFO, "Job Number: {}".format(job_number))
+    logger_run_operations.log(loglevel.INFO, "Output Files: {}, {}".format(stdout_file, stderr_file))
+
+    return [stdout_file, stderr_file], job_number
```
Make return `[stdout_file, stderr_file]` into a tuple

Ovec8hkin replied:
@2gotgrossman Decided I'm not going to do this. It doesn't increase readability or functionality in any way I can see. I think its more obvious what happening if returned as list, since its concatenated into a list.

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
 
-logger = logging.getLogger()
-logger.setLevel(logging.DEBUG)
+def create_process_rsmas_parser():
```
Rename to `run_operations_parser`

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+
+    process_rsmas_cmd = "process_rsmas.py {} {} --submit --update".format(template_file, process_rsmas_options)
+
+    process_rsmas = subprocess.check_output(process_rsmas_cmd, shell=True).decode('utf-8')
```
Determine if subprocess call can be refactored in exchange direct function call

Ovec8hkin replied:
@2gotgrossman Going to leave this for a different diff as it may require a bit of restructuring

gravelcycles replied:
Could you make an issue for this?

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+    
+    if os.path.exists(output_file) and os.path.isfile(output_file):
+
+        base = "{}/{}/{}/".format(os.getenv('OPERATIONS'), 'ERRORS', dset)
```
Use `os.path.join()`

gravelcycles commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+
+def overwrite_stored_date(dset, newest_date):
 
+    new_line = "{}: {}\n".format(dset, newest_date.strftime(DATE_FORMAT))
```
Why not something like this:

```python
with open(file, ‘w’) as f:
	lines = f.readlines()
	new_lines = overwrite_or_append_new_line(lines, new_line)
	f.writelines(lines) # updates all lines in the file
```

Ovec8hkin replied:
Going to leave this for a separate diff as well

gravelcycles commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+    initiate_operations()
+
+    template_files = []
+    # inps.sheet_id is an array of sheets.
```
Why not call it `inps.sheet_ids`?

Ovec8hkin commented on run_operations.py:
```diff
@@ -1,394 +1,293 @@
+        if os.path.isfile(output_files[i]):
+            logger_run_operations.log(loglevel.INFO, "Job #{} of {} complete (output file {})".format(i + 1, len(output_files), output_files[i]))
+            dset = job_to_dset[output_files[i].split("process_rsmas_")[1].split(".")[0]]
+            copy_output_file(output_files[i], dset)
```
Use parameters naming here

Ovec8hkin committed:
Merge remote-tracking branch 'origin/run_operations_new' into run_operations_new

# Conflicts:
#	_processSteps.py
#	process_rsmas.py

Ovec8hkin committed:
Cleaned up unnecesarry import statements

Ovec8hkin committed:
Added documentation to `submit_script` clarifying the return value to
be the job number of the submission

Ovec8hkin committed:
Replace inps parameter with template

Ovec8hkin committed:
Remove extra `generate_template_files_from_dataframe` function

Also added comment to explain return value of `generate_template_files`

Ovec8hkin committed:
Removed extra import statements

Ovec8hkin committed:
Removed extra variable storage from `get_newest_date` and return the
`datetime.strptime` object directly

Ovec8hkin committed:
Removed rename of `generate_template_files` import to `gt` to increase
clarity of import and usage

Ovec8hkin committed:
Removed explicit import of Template in favor of importing all of
`dataset_template` to increase code readability at use cases

Ovec8hkin committed:
Moved all `os.getenv` environment variable calls to global constant and
moved other OPERATIONS dircetories to globals constants

Ovec8hkin committed:
Clarified the opening of STORED_DATE_FILE in `get_last_downloaded_date`
to be for reading only

Ovec8hkin committed:
Renamed to `create_run_operations_parser`

Ovec8hkin committed:
Changed file paths to use os.path.join() for clarity

Ovec8hkin committed:
Changed `inps.sheet_id` to `inps.sheet_ids`

Ovec8hkin committed:
Renamed `copy_output_file` dset parameter to 'dataset' for increase
readability. Also change function call in `run_operations()` to use
parameter names for clarity.

Ovec8hkin committed:
Removed '--update' command line flag from process_rsmas call

Ovec8hkin commented:
@2gotgrossman Provide thoughts on these changes, and then I will test on Pegasus. Once everything is good, I will write documentation for each function and merge.

gravelcycles commented on create_batch.py:
```diff
@@ -12,7 +12,10 @@
+from _process_utilities import send_logger
+from rsmas_logging import loglevel
 
+logger_create_batch = send_logger()
```
Looks like this logging is never used. Can we remove this line and the imports?

Ovec8hkin replied:
Good catch. I think I meant to delete this and missed it.

gravelcycles commented on run_operations.py:
```diff
@@ -51,30 +54,29 @@ def create_process_rsmas_parser():
+    if not os.path.isdir(ERRORS_DIRECTORY):
+        os.mkdir(ERRORS_DIRECTORY)
+
+    if not os.path.exists(OPERATIONS_DIRECTORY + '/stored_date.date'):
```
I think you already have a variable `STORED_DATE_FILE` that is this constant

gravelcycles commented on run_operations.py:
```diff
@@ -51,30 +54,29 @@ def create_process_rsmas_parser():
+        os.mkdir(ERRORS_DIRECTORY)
+
+    if not os.path.exists(OPERATIONS_DIRECTORY + '/stored_date.date'):
+        open(OPERATIONS_DIRECTORY + '/stored_date.date', 'a').close()  # create empty file
```
here too : `STORED_DATE_FILE`

gravelcycles commented on run_operations.py:
```diff
@@ -214,13 +214,14 @@ def run_operations(args):
+    initiate_operations()
 
     template_files = []
     # inps.sheet_id is an array of sheets.
```
Could you update the comment to also handle `inps.sheet_ids`?

gravelcycles commented on run_operations.py:
```diff
@@ -1,394 +1,294 @@
+        if not os.path.exists(base):
+            os.makedirs(base)
+
+        destination = base + str(datetime.now().strftime("%m-%d-%Y")) + os.path.splitext(output_file)[1]
```
It could be cleaner to do a `os.path.join()` to create `destination`. That would also mean you could remove the empty string in L183, which seems hacky

gravelcycles reviewed (commented):
Looks pretty good to me! A few more comments, but other than that, it looks like it's ready for some documentation!!

Could you create issues for the two refactors you are delaying? Here are they are
- https://github.com/geodesymiami/rsmas_isce/pull/108/files#r270637541
- https://github.com/geodesymiami/rsmas_isce/pull/108/files#r270637898

Once documentation is ready, it could be good to get review from @lily-wittle, @mirzaees, and/or @falkamelung to see if there are any rough parts on understanding how the code works. In other words, it would be a good trial run to see if the documentation does its job.

Ovec8hkin committed:
Using STORED_DATE_FILE constant in initiate_operations

Ovec8hkin committed:
Removed logger object from create_batch

Ovec8hkin committed:
Updated inps.sheet_ids comment

Ovec8hkin committed:
Changed destination to use os.path.join()

Ovec8hkin committed:
Added documentation to all functions

gravelcycles commented on create_batch.py:
```diff
@@ -15,8 +15,6 @@
 from _process_utilities import send_logger
 from rsmas_logging import loglevel
```
This and the import above it are unused now that you removed `send_logger()`

gravelcycles commented on run_operations.py:
```diff
@@ -61,7 +63,15 @@ def command_line_parse(args):
 
 
 def initiate_operations():
```
Is there a more appropriate function name? This function name describes *when* it is run but not *what* it does.

Ovec8hkin replied:
Probably, but I don't know what it would be. Lets not worry about it for now.

gravelcycles commented on run_operations.py:
```diff
@@ -70,16 +80,21 @@ def initiate_operations():
 def generate_templates_with_options(csv, dataset, sheet):
-
+    """
+    Uses generate_template_files to generate all active template files based on the provided Google Sheet
```
I'd remove this part: "Uses generate_template_files". That is an implementation detail and is easy to understand from reading the code

gravelcycles commented on run_operations.py:
```diff
@@ -96,7 +111,12 @@ def generate_templates_with_options(csv, dataset, sheet):
 def get_datasets_to_process(template_files, dataset=None):
-
+    """
+    Converts the template files being used in dataset/project names
```
After more thought, function doesn't make much sense to me. There's weird behavior if `template_files` is declared but also `dataset` is declared. I would expect that `template_files` is still used in the function somehow, but it is ignored in this case.

This would be my documentation:
```Extracts dataset names from `dataset` or `template_files` if `dataset` is not declared.```  In other words, you basically have to describe the whole function to explain its functionality. Seems like a pretty shallow abstraction.

gravelcycles replied:
I'd recommend just inlining the function maybe with a single line comment above it.

Ovec8hkin replied:
`template_files` is never used again. Its just to grab all of the dataset names, because I don't want to refer to things by their `template_file` name. I would prefer not to inline it if possible, as I don't see how inlining it, versus leaving it as a function, changes anything.

lily-wittle replied:
Agreed on the documentation `Extracts dataset names...` makes more sense as to what you're doing

gravelcycles commented on run_operations.py:
```diff
@@ -106,7 +126,11 @@ def get_datasets_to_process(template_files, dataset=None):
 def get_newest_data_date(template_file):
-
+    """
+    Uses `ssara_federated_query.py --print` to obtain the most recent image date for a dataset
```
`ssara_federated_query.py` is an implementation detail. No need to expose that in the external documentation. Maybe shift that part of the comment as a single line comment above the part of the code it pertains to?

gravelcycles commented on run_operations.py:
```diff
@@ -1,394 +1,351 @@
+    while i < len(output_files):
+        if os.path.isfile(output_files[i]):
+            logger_run_operations.log(loglevel.INFO, "Job #{} of {} complete (output file {})".format(i + 1, len(output_files), output_files[i]))
+            dset = job_to_dset[output_files[i].split("process_rsmas_")[1].split(".")[0]]
```
Can you provide a comment that gives an example of what this line does? Or can you provide a single line comment explanation?

gravelcycles reviewed (commented):
Made some comments on documentation

Ovec8hkin committed:
Removed unnecessary import statements

Ovec8hkin committed:
Updated docuementation for `generate_templates_with_options` and
`get_newest_data_date`

Ovec8hkin committed:
Added comments about the functionality of the positive branch of the
waiting for files loop

Ovec8hkin committed:
Fixed outfile paths in run_process_rsmas

lily-wittle commented on run_operations.py:
```diff
@@ -1,394 +1,357 @@
+############## DIRECTORY AND FILE CONSTANTS ##############
+OPERATIONS_DIRECTORY = os.getenv('OPERATIONS')
+SCRATCH_DIRECTORY = os.getenv('SCRATCHDIR')
+TEMPLATE_DIRECTORY = OPERATIONS_DIRECTORY + "/TEMPLATES"
```
Should this (and other directories and files) be using `os.path.join()` instead of `/`?

lily-wittle commented on run_operations.py:
```diff
@@ -1,394 +1,357 @@
+
+    ssara_output = subprocess.check_output(ssaraopt_cmd, shell=True)
+
+    newest_data = ssara_output.decode('utf-8').split("\n")[-2]
```
This could be clearer with a comment about what format `ssara_output` is in

lily-wittle commented on run_operations.py:
```diff
@@ -1,394 +1,357 @@
+def copy_output_file(output_file, dataset):
+    """
+    Copies the provided output_file to the $OPERATIONS/ERRORS directory for easy access
+    :param output_file: the output file to be copied (EXAMPLE: process_rsmas_00000001.o)
```
Just a question - why put .o files (which I'm assuming are stdout) into the errors directory?

Ovec8hkin replied:
You make a good point. Should probably rename the directory to `OUTPUTS`

lily-wittle commented on run_operations.py:
```diff
@@ -1,394 +1,357 @@
+
+            #  Exit and don't overwrite date file if process_rsmas.py throws and error
+            try:
+                # Submit processing job and running processing routine vis process_rsmas.py
```
Nitpick: "via"?

lily-wittle reviewed (commented):
Overall looks good to me! I put a few comments on things that I didn't follow very well, but otherwise pretty clear what each function is doing

Ovec8hkin committed:
Updates

Ovec8hkin committed:
Changed file and directory and constants to use os.path.join

Ovec8hkin committed:
Updated documentation for `get_datasets_to_process`

Ovec8hkin committed:
Added comment about format of ssara_output

lily-wittle commented on run_operations.py:
```diff
@@ -16,10 +16,10 @@
-LOGS_DIRECTORY = OPERATIONS_DIRECTORY + "/LOGS"
-ERRORS_DIRECTORY = OPERATIONS_DIRECTORY + "/ERRORS"
-STORED_DATE_FILE = OPERATIONS_DIRECTORY + "/stored_date.date"
+TEMPLATE_DIRECTORY = os.path.join(OPERATIONS_DIRECTORY, "/TEMPLATES")
```
I think the `/` needs to be removed - won't this discard `OPERATIONS_DIRECTORY` since `/TEMPLATES` is an absolute path?

Ovec8hkin replied:
Good catch

Ovec8hkin committed:
Removed leading slack on concatenated directory names

Ovec8hkin committed:
Changed 'run_operations.log' file defition to use os.path.join
